Make iptables and ipvs modes of kube-proxy MASQUERADE --random-fully if possible

MikeSpreitzer opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. Follow the instructions for writing a release note: https://git.k8s.io/community/contributors/guide/release-notes.md
6. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

**What type of PR is this?**
> Uncomment only one ` /kind <>` line, hit enter to put that in a new line, and remove leading whitespaces from that line:
>
> /kind api-change

/kind bug

> /kind cleanup
> /kind design
> /kind documentation
> /kind failing-test
> /kind feature
> /kind flake

**What this PR does / why we need it**:
This PR makes kube-proxy in iptables or ipvs mode include `--random-fully` in the iptables rule it emits to `-j MASQUERADE`, if the local version of iptables supports that additional flag.  This greatly reduces the problems caused by a Linux kernel bug that bites when multiple flows are started at the same time.  Due to this kernel bug the flows can get mapped to the same port, and all but one of them will suffer packet drops.  This flag makes the port choice fully random, greatly reducing the probability of this port collision and consequent packet drop.

**Which issue(s) this PR fixes**:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Fixes #76699

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".
-->
```release-note
NONE
```

MikeSpreitzer commented:
@kubernetes/sig-network-bugs

MikeSpreitzer commented:
/remove-label needs-sig

MikeSpreitzer commented:
/woof

cmluciano commented:
/assign @cmluciano

cmluciano commented on pkg/proxy/iptables/proxier.go:
```diff
@@ -772,12 +788,19 @@ func (proxier *Proxier) syncProxyRules() {
 		"-j", "MASQUERADE",
-	}...)
+	}
+	if proxier.masqueradeRandomFully {
```
Please add an e2e to be sure this is parsed correctly when this feature is enabled

MikeSpreitzer replied:
The existing code will log an error if there is any problem with any rule.

Are there already e2e test(s) that exercise kube-proxy and check that this error is not logged?  If so, then this is already covered.  If not, is it fair to ask for this PR to introduce that?

cmluciano replied:
That may be the case, could you reference the kube-proxy test that leverages this generated code for the test cases

I'm mainly trying to find unit test that just validates the expected lines added by this flag or an e2e that enables the flag and ensures the baseline iptables functionality is correct. I can imagine that having the e2e for the scale tests would be useful to see if we can replicate the issue without this flag enabled.

danwinship replied:
The code isn't that confusing, though maybe it would be nice to just have `pkg/util/iptables/` just check the version once and export a bunch of variables like `SupportsCheck` and `SupportsRandomFully` for everyone else to use.

MikeSpreitzer replied:
On the original point, of testing:

1. I added a unit test that the generated `-j MASQUERADE` includes `--random-fully` _if_ the iptables version supports it.

2. The existing e2e tests surely test that all the generated rules parse, for the version of iptables where the tests run.  If any rule does not parse then none are installed, which causes massive damage --- for example, no client will be able to reach the `kubernetes` Service.

MikeSpreitzer commented on pkg/proxy/iptables/proxier.go:
```diff
@@ -308,6 +313,16 @@ func NewProxier(ipt utiliptables.Interface,
+	masqueradeRandomFully := false
+	iptVersionString, err := ipt.GetVersion()
+	if err == nil {
+		iptVersion, err := utilversion.ParseGeneric(iptVersionString)
```
I am having trouble understanding the words here; this line invokes a different function than line 317.  Can you please elaborate on the problem?

cmluciano replied:
I got my lines confused. I meant to comment on L318-328 where we run ParseGeneric a few times in different nested ifs. Can we consolidate these lines and utilize something like [version.Compare](https://github.com/kubernetes/apimachinery/blob/master/pkg/util/version/version.go#L308). It might just be me but I got a little lost walking through all of these nested conditions. I also had to reread the final comparison on L326 a few times before I understood that we are setting that flag to true based on the other conditions. 

Perhaps @dcbw or @danwinship will have other thoughts on if it looks ok to them.

MikeSpreitzer replied:
@cmluciano : in lines 318--328 there are two different strings parsed with `utilversion.ParseGeneric`.  We coud remove one of those by using `version.Compare` --- which does a parse internally --- instead of `version.AtLeast`.

MikeSpreitzer replied:
@danwinship : it looks like https://github.com/kubernetes/kubernetes/pull/78547#discussion_r296884083 was intended to be attached here, right?

I am a little surprised that `pkg/util/iptables` does not cache the version string.  Is that deliberate preparation for coping with live kernel patching?  Should we continue to eschew caching the version string and things derived from it for that reason?

There is no direct external consumer for the "supports check" bit, so "SupportsRandomFully" (data bit or query function) would be the first of its kind.

danwinship replied:
> @danwinship : it looks like [#78547 (comment)](https://github.com/kubernetes/kubernetes/pull/78547#discussion_r296884083) was intended to be attached here, right?

oops. yes

> I am a little surprised that `pkg/util/iptables` does not cache the version string. Is that deliberate preparation for coping with live kernel patching?

Nothing so planned. Originally it was only needed in one place. Later whoever added the second usage must have felt it would be easier to just call the "get version" function again rather than cache it. it's not like it takes that long to run "`iptables --version`", and you only end up doing it a few times anyway...

> There is no direct external consumer for the "supports check" bit, so "SupportsRandomFully" (data bit or query function) would be the first of its kind.

There are external consumers (like `pkg/proxy/iptables/proxier.go`), they just have to implement it by getting the iptables version and manually comparing it against `utiliptables.MinCheckVersion`...

MikeSpreitzer replied:
I have refactored, consolidating the version string parsing and comparison logic.

MikeSpreitzer commented on pkg/proxy/iptables/proxier.go:
```diff
@@ -308,6 +313,16 @@ func NewProxier(ipt utiliptables.Interface,
 	}
 
+	masqueradeRandomFully := false
+	iptVersionString, err := ipt.GetVersion()
```
yes.  Done.

MikeSpreitzer commented:
/retest

MikeSpreitzer commented:
/test pull-kubernetes-verify

MikeSpreitzer commented:
/test pull-kubernetes-verify

MikeSpreitzer commented:
/test pull-kubernetes-verify

MikeSpreitzer commented:
/retest

cmluciano commented:
Can you run ./hack/update-bazel.sh

MikeSpreitzer commented:
@danwinship , @cmluciano : I think I have addressed all the comments.  Look good?

cmluciano commented:
thanks for consolidating and adding the unit test

/lgtm

MikeSpreitzer commented:
/assign @dcbw

vllry commented:
LGTM for what it's worth

/sig network

danwinship commented on pkg/util/iptables/iptables.go:
```diff
@@ -129,6 +129,39 @@ const WaitSecondsMinVersion = "1.4.22"
+
+type TestOutcome struct {
+	Answer bool
+	Err    error
```
er... no. Just use a multiple-value return from `VersionStringIsAtLeast`

MikeSpreitzer replied:
What is bad about defining a return type?  It has the advantage that a caller that wants just one member can make the call and select the member all in one expression; a multiple-value return forces the caller to write something more long-winded.

MikeSpreitzer replied:
At https://github.com/kubernetes/kubernetes/blob/655de9c860f1658ec86ded978dc99de8038e5c17/pkg/proxy/iptables/proxier_test.go#L376 you can see an example of the advantage of returning one struct rather than multiple values.

danwinship replied:
It's non-go-like. But anyway, this is why I was saying we should do all the version checks once at the beginning and then just have "SupportsCheck()", "SupportsRandomFully()" etc methods. Then you don't need to worry about returning errors, because that will already have been dealt with a long time before.

#80368 refactors the utiliptables version checking and gets rid of `GetVersion()`. Originally it added a `SupportsCheck()` to replace it before we realized we didn't even need that. But you could set a `runner.supportsRandomFully` variable when first creating the object, and have a `SupportsRandomFully()` method that just returns that. (This also keeps all of the knowledge of specific version numbers inside utiliptables rather than having its consumers need to know them.)

MikeSpreitzer replied:
I see.  It looks like #80368 is establishing some groundwork to build on.

MikeSpreitzer committed:
Made IPVS and iptables modes of kube-proxy fully randomize masquerading if possible

Work around Linux kernel bug that sometimes causes multiple flows to
get mapped to the same IP:PORT and consequently some suffer packet
drops.

Also made the same update in kubelet.

Also added cross-pointers between the two bodies of code, in comments.

Some day we should eliminate the duplicate code.  But today is not
that day.

danwinship reviewed (changes requested):
Mostly looks good

danwinship commented:
/lgtm
but we're going to need a kubelet owner now too... sorry

MikeSpreitzer commented:
/assign @thockin

thockin reviewed (commented):
Thanks!

/lgtm
/approve

MikeSpreitzer commented:
How are those job failures even possible?  The same commit passed those jobs earlier.  And the line numbers in the complaints do not match the line numbers in the code!

MikeSpreitzer commented:
/retest

MikeSpreitzer commented:
@danwinship : I had to rebase, so this needs LGTM again.

danwinship commented:
/lgtm

MikeSpreitzer commented:
/priority important-soon

alvaroaleman commented:
@MikeSpreitzer Is it possible to file an exception request for this to still get it into 1.16? IMHO this is a pretty important fix.

MikeSpreitzer commented:
I am not sure.  Where is that process defined?

MikeSpreitzer commented:
Note also that this PR does not have a KEP, I thought it was too small a thing for that.

cblecker commented:
What is the current minimum supported version of iptables? This change would require it to be at least 1.6.2, as that is when this feature/flag was introduced to iptables.

MikeSpreitzer commented:
@cblecker : this PR does not add any requirement.  As suggested in the title, the `--random-fully` flag is only issued if it is supported.

cblecker commented:
Ah yes, I see that now.

Should we make this bit of magic discoverable in some way? I don’t see any log messages (even at debug) if it doesn’t enable random-fully.. I see this potentially making things difficult from the sysadmin perspective.. you magically just get a different functionality based on the version of iptables you’re running.

MikeSpreitzer commented:
Dang, I suppose a revision is needed.  A log entry seems the most direct approach.  What V level would you recomend?  As it is once per startup, I think something prominent like 2 or 3 would be appropriate.

cblecker commented:
If it's run once, I'd be okay with a V(2). If this iptables feature is fixing a bug, as a sysadmin, I'd want to know if we are falling back to not enabling it because the iptables version is too old.

MikeSpreitzer commented:
Hmm, now that I think about it, there are two cases: when the feature is supported and when it is not.  How about a V(3) when good and a warning when bad?

cblecker commented:
That works too. Less concerned about logs for the "good" state (if we log this, V(3) sounds right to me), but having something prominent for the not working state I think is important.

The question on an V(2) info vs a warning for me would be: warning should be a call to action, that even though things are working, the sysadmin *needs* to do something. A V(2) info is more just like.. here's a bit of info if you end up going looking for it. Is the bug bad enough that we need that call to action?

MikeSpreitzer commented:
OK, right, I think the case of low iptables version should be a matter of info rather than a call to action.  So V(2) for bad, V(3) for good.

MikeSpreitzer commented:
OK, I just force-pushed that revision.

cblecker commented:
/lgtm
/approve

@lachie83 My concerns have been fully addressed.

MikeSpreitzer commented:
@cblecker : thanks!

MikeSpreitzer commented:
/retest

MikeSpreitzer commented:
Dang!  I had to push another revision to fix a mis-spelling.

MikeSpreitzer commented:
@cblecker

cblecker commented:
/lgtm

MikeSpreitzer commented:
thanks again!

lachie83 commented:
Does this fix address a bug that was introduced by an enhancement in 1.16? If not, are you planning on getting this cherry-picked in the current supported releases?

MikeSpreitzer commented:
@lachie83 : the bug that this fixes was reported in April, but has been around longer than that.  This is a work-around for a Linux kernel deficiency or bug that has been there forever (as far as I know).  Yes, cherry-picking this to supported releases would be good.  Regarding risks: this is the usual work-around for the kernel issue, and I am not aware of any reported problems with this work-around.

MikeSpreitzer commented:
@lachie83, another thought regarding risks: I am not sure how `-j MASQUERADE --random-fully` behaves when the kernel is running low on available ports; there could be a risk there.

dannyk81 commented:
@lachie83 @MikeSpreitzer I'd also comment that this flag has become standard (for a while now) in recent releases of CNI plugins such as Calico and Flannel that create iptables rules.

Calico: https://github.com/projectcalico/felix/pull/1901
Flannel: https://github.com/coreos/flannel/pull/1040

MikeSpreitzer commented:
Also FYI: Here is an explanation of the problem and how `--random-fully` works around it: https://tech.xing.com/a-reason-for-unexplained-connection-timeouts-on-kubernetes-docker-abd041cf7e02 .  The code cited in reference (3) includes the loop at https://github.com/torvalds/linux/blob/24de3d377539e384621c5b8f8f8d8d01852dddc8/net/netfilter/nf_nat_proto_common.c#L87-L94 , which answers the question of how `--random-fully` reacts to high utilization of the port range: no worse than the alternatives.  The only difference is where the search starts.

MikeSpreitzer commented:
That same code structure is in the latest kernel in the 4.X line.
The 5.X line is different.

lachie83 commented:
Thanks for the updates. Can we get issue https://github.com/kubernetes/kubernetes/issues/76699 appropriately labelled with the correct priority from sig-network so that I can evaluate the urgency of getting this milestoned in 1.16?

MikeSpreitzer commented:
@lachie83 : can you please elaborate on what that means / how to get it done?

MikeSpreitzer commented:
BTW, on further investigation: it looks like in the 5.X kernel they have made this randomization the normal behavior; the `--random-fully` option is still accepted but no longer makes a difference.

lachie83 commented:
/milestone v1.16
